### PR TITLE
Transcendental tangent planes utilities

### PR DESCRIPTION
**Pull request by ajreynol**
Work on #1168 (improved support for transcendental functions, requested by GE).

This commit:
- Adds necessary support for a few utilities that are necessary for #1168, including derivatives, Taylor series, Taylor error functions, and concavity regions.
- Fixes a bug in isExtfReduced which led to unnecessary lemmas involving transcendental functions.
- Improves the initial refinement lemma scheme for transcendental functions.
- Adds some missing documentation in nonlinear_extension.h regarding some of the prominent functions and those needed to understand the forthcoming commits.
- References two new open issues for further fixes/documentation to non-linear #1284 and #1287.

This is part 2/3 for merging https://github.com/ajreynol/CVC4/tree/nltTfTangentPlanes-v2

### CODE REVIEW

File: src/theory/arith/nonlinear_extension.h [hunk 1]
```diff
@@ -176,13 +361,29 @@ class NonlinearExtension {
...
-  // utilities
+  /** commonly used terms */
   Node d_zero;
   Node d_one;
   Node d_neg_one;
   Node d_true;
   Node d_false;
-
+  /** PI 
+   * Note that PI is a (symbolic, non-constant)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 2]
```diff
@@ -197,7 +398,10 @@ class NonlinearExtension {
   std::vector<Node> d_constraints;
 
   // model values/orderings
-  // model values
+  /** cache of model values
+   * Stores the the concrete/abstract model values
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 3]
```diff
@@ -246,29 +451,260 @@ class NonlinearExtension {
...
+   * transcendental functions. This is the set
+   * "get-previous-secant-points" in "Satisfiability
+   * Modulo Transcendental Functions via Incremental
+   * Linearization" by Cimatti et al., CADE 2017, for
+   * each transcendental function application.
+   */
+  std::unordered_map<Node, std::vector<Node>, NodeHashFunction> d_secant_points;
+
+  /** get Taylor series of degree n for function fa centered around point fa[0].
+   * Return value is ( P_{n,f(a)}( x ), R_{n+1,f(a)}( x ) ) where
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 4]
```diff
@@ -246,29 +451,260 @@ class NonlinearExtension {
...
+  *   region #4 is -pi < x <= -pi/2
+  *   region #5 is -infty < x <= -pi (this is an invalid region)
+  * All regions not listed above, as well as regions 0 and 5
+  * for SINE are "invalid". We only process applications
+  * of transcendental functions whose arguments have model
+  * values that reside in valid regions.
+  */
+  std::unordered_map<Node, int, NodeHashFunction> d_tf_region;
+  /** get monotonicity direction
+  * Returns whether the slope is positive (+1) or negative(-1)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 5]
```diff
@@ -246,29 +451,260 @@ class NonlinearExtension {
...
+  * in region of transcendental function with kind k.
+  * Returns 0 if region is invalid.
+  */
+  int regionToMonotonicityDir(Kind k, int region);
+  /** get concavity
+  * Returns whether we are concave (+1) or convex (-1)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 6]
```diff
@@ -246,29 +451,260 @@ class NonlinearExtension {
...
+  * in region of transcendental function with kind k,
+  * where region is defined above.
+  * Returns 0 if region is invalid.
+  */
+  int regionToConcavity(Kind k, int region);
+  /** region to lower bound 
+   * Returns the term corresponding to the lower
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 7]
```diff
@@ -246,29 +451,260 @@ class NonlinearExtension {
...
+   * bound of the region of transcendental function
+   * with kind k. Returns Node::null if the region 
+   * is invalid, or there is no lower bound for the 
+   * region.
+   */
+  Node regionToLowerBound(Kind k, int region);
+  /** region to upper bound
+   * Returns the term corresponding to the upper
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 8]
```diff
@@ -246,29 +451,260 @@ class NonlinearExtension {
...
+   * bound of the region of transcendental function
+   * with kind k. Returns Node::null if the region 
+   * is invalid, or there is no upper bound for the 
+   * region.
+   */
+  Node regionToUpperBound(Kind k, int region);
+  /** Get derivative.
+  * Returns d/dx n. Supports cases of n
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ditto

File: src/theory/arith/nonlinear_extension.h [hunk 9]
```diff
@@ -246,29 +451,260 @@ class NonlinearExtension {
...
+  *  ( y>=0 ^ s <= x*z ^ x*y <= t ) => y*s <= z*t
+  *  ...where s <= x*z and x*y <= t are constraints 
+  *     that occur in the current context.
+  */
   std::vector<Node> checkMonomialInferResBounds();
-  
+
+  /** check tangent planes
+  *
+  * Returns a set of valid theory lemmas, based on a
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: a --> an

File: src/theory/arith/nonlinear_extension.cpp [hunk 10]
```diff
@@ -663,31 +669,48 @@ bool NonlinearExtension::getCurrentSubstitution(
...
-        if (exp[i][j].getKind() == kind::EQUAL) {
-          eqs.push_back(exp[i][j]);
+  if (on.getKind() == kind::NONLINEAR_MULT)
+  {
+    Trace("nl-ext-zero-exp") << "Infer zero : " << on << " == " << n
+                             << std::endl;
+    // minimize explanation if a substitution+rewrite results in zero
+    const std::set<Node> vars(on.begin(), on.end());
+
+    for (unsigned i = 0; i < exp.size(); i++)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: no ``size()`` in condition.

File: src/theory/arith/nonlinear_extension.cpp [hunk 11]
```diff
@@ -663,31 +669,48 @@ bool NonlinearExtension::getCurrentSubstitution(
...
+      Trace("nl-ext-zero-exp") << "  exp[" << i << "] = " << exp[i]
+                               << std::endl;
+      std::vector<Node> eqs;
+      if (exp[i].getKind() == kind::EQUAL)
+      {
+        eqs.push_back(exp[i]);
+      }
+      else if (exp[i].getKind() == kind::AND)
+      {
+        for (unsigned j = 0; j < exp[i].getNumChildren(); j++)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: for-each

File: src/theory/arith/nonlinear_extension.cpp [hunk 12]
```diff
@@ -1608,13 +1636,20 @@ int NonlinearExtension::compare_value(Node i, Node j,
...
-  Assert(orderType >= 0 && orderType <= 3);
-  unsigned mindex = orderType <= 1 ? 0 : 1;
-  std::map<Node, Node>::const_iterator iti = d_mv[mindex].find(i);
-  Assert(iti != d_mv[mindex].end());
-  return iti->second;
+  if (i.isConst())
+  {
+    return i;
+  }
+  else
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: no-else-after-return

File: src/theory/arith/nonlinear_extension.cpp [hunk 13]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+      return 1;
+    }
+  }
+  else if (k == kind::SINE)
+  {
+    if (region == 1 || region == 4)
+    {
+      return -1;
+    }
+    else if (region == 2 || region == 3)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 14]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+      return 1;
+    }
+  }
+  else if (k == kind::SINE)
+  {
+    if (region == 1 || region == 2)
+    {
+      return -1;
+    }
+    else if (region == 3 || region == 4)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 15]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+
+Node NonlinearExtension::regionToLowerBound(Kind k, int region)
+{
+  if (k == kind::SINE)
+  {
+    if (region == 1)
+    {
+      return d_pi_2;
+    }
+    else if (region == 2)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 16]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+    {
+      return d_zero;
+    }
+    else if (region == 3)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 17]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+    {
+      return d_pi_neg_2;
+    }
+    else if (region == 4)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 18]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+
+Node NonlinearExtension::regionToUpperBound(Kind k, int region)
+{
+  if (k == kind::SINE)
+  {
+    if (region == 1)
+    {
+      return d_pi;
+    }
+    else if (region == 2)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 19]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+    {
+      return d_pi_2;
+    }
+    else if (region == 3)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 20]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+    {
+      return d_zero;
+    }
+    else if (region == 4)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 21]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+{
+  Assert(x.isVar());
+  // only handle the cases of the taylor expansion of d
+  if (n.getKind() == kind::EXPONENTIAL)
+  {
+    if (n[0] == x)
+    {
+      return n;
+    }
+    else
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 22]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+  else if (n.getKind() == kind::SINE)
+  {
+    if (n[0] == x)
+    {
+      Node na = NodeManager::currentNM()->mkNode(kind::MINUS, d_pi_2, n[0]);
+      Node ret = NodeManager::currentNM()->mkNode(kind::SINE, na);
+      ret = Rewriter::rewrite(ret);
+      return ret;
+    }
+    else
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 23]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+  {
+    std::vector<Node> dchildren;
+    for (unsigned i = 0; i < n.getNumChildren(); i++)
+    {
+      // PLUS is flattened in rewriter, recursion depth is bounded by 1
+      Node dc = getDerivative(n[i], x);
+      if (dc.isNull())
+      {
+        return dc;
       }else{
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 24]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
-        i_derv = tf;
+        dchildren.push_back(dc);
       }
-      if( i_derv_status>=2 ){
-        i_derv = NodeManager::currentNM()->mkNode( kind::MINUS, d_zero, i_derv );
+    }
+    return NodeManager::currentNM()->mkNode(kind::PLUS, dchildren);
+  }
+  else if (n.getKind() == kind::MULT)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 25]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+    {
+      return NodeManager::currentNM()->mkNode(kind::MULT, n[0], dc);
+    }
+  }
+  else if (n.getKind() == kind::NONLINEAR_MULT)
+  {
+    unsigned xcount = 0;
+    std::vector<Node> children;
+    unsigned xindex = 0;
+    for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: no ``getNumChildren()`` in condition as well?

File: src/theory/arith/nonlinear_extension.cpp [hunk 26]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
       }
-      i_derv = Rewriter::rewrite( i_derv );
-      i_derv_status = i_derv_status==3 ? 0 : i_derv_status+1;
+      children.push_back(n[i]);
+    }
+    if (xcount == 0)
+    {
+      return d_zero;
+    }
+    else
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: no-else-after-return

File: src/theory/arith/nonlinear_extension.cpp [hunk 27]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
-    Node curr = NodeManager::currentNM()->mkNode( kind::MULT, 
-                  NodeManager::currentNM()->mkNode( kind::DIVISION, i_derv, i_fact ), i_exp );
-    sum.push_back( curr );
-    i_fact = Rewriter::rewrite( NodeManager::currentNM()->mkNode( kind::MULT, NodeManager::currentNM()->mkConst( Rational( counter ) ) ) );
-    i_exp = Rewriter::rewrite( NodeManager::currentNM()->mkNode( kind::MULT, i_exp_base, i_exp ) );
-  }while( counter<n );
-  return sum.size()==1 ? sum[0] : NodeManager::currentNM()->mkNode( kind::PLUS, sum );
+    return NodeManager::currentNM()->mkNode(kind::MULT, children);
+  }
+  else if (n.isVar())
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 28]
```diff
@@ -2895,53 +2910,311 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+  {
+    return n == x ? d_one : d_zero;
+  }
+  else if (n.isConst())
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

File: src/theory/arith/nonlinear_extension.cpp (hunk 13 above)
Comment (ajreynol) [reply src/theory/arith/nonlinear_extension.cpp]: I think since this else doesn't add any space, I'd prefer keeping it

Comment (ajreynol) [reply src/theory/arith/nonlinear_extension.cpp]: same goes for those below

File: src/theory/arith/nonlinear_extension.h [hunk 29]
```diff
@@ -42,25 +43,114 @@ namespace arith {
...
   bool getCurrentSubstitution(int effort, const std::vector<Node>& vars,
                               std::vector<Node>& subs,
                               std::map<Node, std::vector<Node> >& exp);
-
+  /** Is the term n in reduced form?
+   * 
+   * Used for context-dependent simplification.
+   * 
+   * effort is an identifier indicating the stage where
+   *   we are performing context-dependent simplification,
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/arith/nonlinear_extension.h]: I've improved this in my latest commit (indenting more)

**Pull request merged**